File: fabric-mge-backend/apps/search/Query.py
from apps.search.models.Query import Query
from django.http import JsonResponse
import json
from apps.storage.loggger import logger
from apps.storage.init import mpt
from apps.storage.utils import MptStrToArray


def Summary(categories, data, q_id):
    for category in categories:
        if data['category'] == category['name']:
            if category['templates'] is not None:
                for temp in category['templates']:
                    if temp['name'] == data['template']['title']:
                        category['count_at_least'] = category['count_at_least'] + 1
                        temp['count_at_least'] = temp['count_at_least'] + 1
                        return categories
                category['templates'].append({  # 此模板没有出现过
                    'count_at_least': 1,
                    'id': data['template']['id'],
                    'name': data['template']['title'],
                    'download': False,
                    'url': '/api/v2/search/query/' + str(q_id) + '/' + str(data['template']['id'])
                })
                return categories
    categories.append({  # 此category没有出现过
        'count_at_least': 1,
        'id': data['category'],
        'name': data['category'],
        'templates': [
            {
                'count_at_least': 1,
                'id': data['template']['id'],
                'name': data['template']['title'],
                'download': False,
                'url': '/api/v2/search/query/' + str(q_id) + '/' + str(data['template']['id'])
            }
        ]
    })
    return categories


def search(text, q_id):
    data_origin = mpt.get(text.encode('utf-8'))
    data = MptStrToArray(data_origin.data)
    data_ = []
    summary = {
        'category': [],
        'keywords': [],
        'realname': []
    }
    for item in data:
        data_.append({
            'download': item['downloads'],
            'score': item['score'],
            'data': {
                'abstract': item['abstract'],
                'add_time': item['add_time'],
                'category': item['category'],
                'category_name': item['category'],
                'downloads': item['downloads'],
                'external_link': None,
                'id': item['id'],
                'methods': item['methods'],
                'project': item['project'],
                'realname': item['author'],
                'score': item['score'],
                'source': item['source'],
                'subject': item['subject'],
                'template': item['template']['id'],
                'template_name': item['template']['title'],
                'title': item['title'],
                'user': item['author'],
                'views': item['views']
            }
        })
        categories = summary['category']
        summary['category'] = Summary(categories, item, q_id)

    return data_, summary


def searchAPI(request):  # 不带page的search
    logger(request, 'Query')
    data = json.loads(request.body.decode().replace("'", "\""))
    q = data.get('q')
    text = q.get('text')
    query = Query(
        value={'text': text},
    )
    res, summary = search(text, query.id)
    query.data = res
    query.summary = summary
    query.total = len(res)
    query.save()
    return JsonResponse({
        'id': query.id,
        'q': {
            'data': query.data,
            'summary': query.summary,
            'total': query.total,
            'value': query.value,
            'page': 1,
            'page_size': 20
        },
        'username': None
    })


def searchAPITemplate(request, q_id, t_id):
    logger(request, 'Query')
    query = Query.objects(id=q_id)[0]
    data = query.data
    res = []
    for item_ in data:
        item = item_['data']
        if item['template'] == t_id:
            res.append({
                'data': {
                    'abstract': item['abstract'],
                    'add_time': item['add_time'],
                    'category': item['category'],
                    'category_name': item['category_name'],
                    'downloads': item['downloads'],
                    'external_link': None,
                    'id': item['id'],
                    'methods': item['methods'],
                    'project': item['project'],
                    'realname': item['realname'],
                    'score': item['score'],
                    'source': item['source'],
                    'subject': item['subject'],
                    'template': item['template'],
                    'template_name': item['template_name'],
                    'title': item['title'],
                    'user': item['realname'],
                    'views': item['views']
                },
                'download': 0,
                'score': 0
            })
    return JsonResponse({
        'data': res,
        'total': len(res),
        'page': 1,
        'pageSize': 20,
    })


def searchAPIPage(request, q_id):
    logger(request, 'Query')
    query = Query.objects(id=int(q_id))[0]
    return JsonResponse({
        'id': query.id,
        'q': {
            'data': query.data,
            'summary': query.summary,
            'total': query.total,
            'value': query.value,
            'page': 1,
            'page_size': 20
        },
        'username': None
    })

File: fabric-frontend/apis/search/v2/Query.ts
import { Search } from '../../define/search';
import JsonApiFetch, { RestApiFetch,RestApiFetchV2 } from '../../Fetch';
import Urls from '../../Urls';
import { JoinParameters } from '../../../utils/JoinParameters';


export interface DataQuery {
  tid: number;
  q: Search.Condition.Any;
}

export interface QueryConfig {
  meta?: Search.Condition.Any;
  text?: string;
  data?: DataQuery[];
  sort?: { [x: string]: 'desc' | 'asc' };
}

export interface Summary {
  category: QuerySummaryItem[];
  keywords: Array<{doc_count: number, key: string}>;
  realname: Array<{doc_count: number, key: string}>;
}

export interface Query {
  id: number;
  username: string;
  q: {
    value: QueryConfig;
    summary: Summary;
    page: number;
    page_size: number;
    total: number;
    data: QueryResultItem[];
  }
  clientVO: any;
}

export namespace Query {
  export function copy(q: Query): Query {
    return {
      id: q.id,
      username: q.username,
      clientVO: q.clientVO,
      q: {
        value: { ...q.q.value},
        summary: {
          category: [...q.q.summary.category],
          realname: [...q.q.summary.realname],
          keywords: [...q.q.summary.keywords],
        },
        page: q.q.page,
        page_size: q.q.page_size,
        total: q.q.total,
        data: [...q.q.data]
      }
    }
  }
}

export const TemplateList = [2346, 2345, 2363, 2348, 2349, 2350, 2364, 2351, 2365, 2369, 2432,
2370, 2371, 2352, 2353, 2354, 2355, 2356, 2357, 2358, 2359, 2360, 2425, 2361, 2220,
  2219, 2216, 2217, 2218, 2426, 2383, 2366, 2429, 2362, 2430, 2428, 2427, 2431, 2472,
2477, 2476, 2525, 2474, 2502, 2473];

// 模板ID若为TemplateList中的，需要向南大通用接口发起检索
export async function CreateConvertQuery(config: QueryConfig){
  return RestApiFetch(Urls.api_v2_search.convert_query_list, 'POST', {q: config});
}

export async function CreateQuery(config: QueryConfig) {
  // return RestApiFetch<Query>(Urls.api_v2_search.query_list, 'POST', { q: config });
  return RestApiFetch<Query>('http://localhost:9001' + Urls.api_v2_search.query_list, 'POST', { q: config });
}

export async function PatchQuery(id: number, config: QueryConfig) {
  return RestApiFetch<Query>(Urls.api_v2_search.query_total(String(id)), 'PATCH', { q: config })
}

export async function PutQuery(id: number, config: QueryConfig) {
  return RestApiFetch<Query>(Urls.api_v2_search.query_total(String(id)), 'PUT', { q: config })
}

export interface QueryTemplateItem {
  id: number;
  name: string;
  url: string;
  count_at_least: number;
  download: boolean;
}

export interface QuerySummaryItem {
  id: number;
  name: string;
  count_at_least: number;
  templates: QueryTemplateItem[];
}

export interface QueryResultData {
  title: string;
  methods: string[];
  project: string;
  subject: string;
  id: number;
  category: number;
  category_name: string;
  template: number;
  template_name: string;
  user: string;
  realname: string;
  importing: boolean;
  score: number;
  downloads: number;
  views: number;
  add_time: string;
  is_public: boolean;
  review_state: 'pending' | 'approved' | 'disapproved',
  summary: string;
  abstract: string;
  doi: string;
  source: string;
  reference: string;
}

export interface QueryResultItem {
  data: QueryResultData;
  download: boolean;
  score: number;
}

export interface GetQueryParam {
  page?: number;
  page_size?: number;
  // base64
  return_meta?: string[] | string;
  // base64
  return_data?: string[] | string;
}

export async function GetQuery(id: string, param: GetQueryParam) {
  let url = JoinParameters(Urls.api_v2_search.query_total(id), param);
  url = 'http://localhost:9001/api/v2/search/query/?q_id=' + id;
  return RestApiFetch<Query>(url, 'GET');
}

export interface DataDetailItem {
  data: {
    // 数据的id
    id: number | string;
    // 模板的id
    template: number;
    // 数据的内容
    // 只包含必要的字段，比如只有容器A下面的B字段，那么content就是{ 'A': { 'B': 123 } }
    content: any;
    [x: string]: string | number;
  }

  // 是否在下载列表中
  download: boolean;
  // 数据的评分
  score: number;
}

export interface QueryTemplateDetail {
  page: number;
  page_size: number;
  total: number;
  data: DataDetailItem[];
}

export async function GetQueryTemplate(id: string, template_id: number, param: GetQueryParam) {
  let url = Urls.api_v2_search.query_template_detail(id, template_id);
  url = JoinParameters(url, param);
  url = 'http://localhost:9001/api/v2/search/query/?q_id=' + id + '&t_id=' + template_id;
  return RestApiFetch<QueryTemplateDetail>(url, 'GET');
}

export async function GetQueryDataList(id:number,param:GetQueryParam) {    //for川大
  let urlmat =  window.location.href.split('/')
  let url = Urls.api_v2_search.query_datails(id,urlmat[urlmat.length-1],);
  url = JoinParameters(url, param);
  return RestApiFetchV2<QueryTemplateDetail>(url,'GET')
}

export interface QueryDownloadItem {
    title: string;
    id: number;
    template: number;
  }

export interface QueryDownload {
  download: {
    data: {
      include: QueryDownloadItem[],
      exclude: QueryDownloadItem[],
    },
    template: {
      include: QueryDownloadItem[],
      exclude: QueryDownloadItem[],
    }
  }
}

export async function GetQueryDownload(id: number) {
  return RestApiFetch<QueryDownload>(Urls.api_v2_search.query_download(id));
}

export interface PartialQueryDownload {
  download: {
    data?: {
      include?: number[],
      exclude?: number[],
    },
    template?: {
      include?: number[],
      exclude?: number[],
    }
  }

}

export async function PatchQueryDownload(id: number, data: PartialQueryDownload) {
  return RestApiFetch<QueryDownload>(Urls.api_v2_search.query_download(id), 'PATCH', data);
}

export async function AddDataToQueryDownload(queryID: number, dataID: number) {
  return PatchQueryDownload(queryID, { download: { data: { include: [dataID] } } });
}

export async function RemoveDataFromQueryDownload(queryID: number, dataID: number) {
  return PatchQueryDownload(queryID, { download: { data: { exclude: [dataID] } } });
}


export async function PutQueryDownload(id: number, data: QueryDownload) {
  return RestApiFetch<QueryDownload>(Urls.api_v2_search.query_download(id), 'PUT', data);
}

export async function CreateFullTextQuery(text: string) {
  return CreateQuery({ text });
}

export async function CreateMetaDataQuery(type: Search.MetaType, value: string) {
  return CreateQuery({
    meta: { and: [{ field: type, val: value, op: Search.Operator.StrContain }] }
  });
}
